File: mobile/lib/features/client_profile/domain/entities/client_profile.dart
import '../../../../core/models/review.dart';

/// Provider associated with a completed client project entry.
///
/// Kept intentionally narrow — only the fields the public client
/// profile surfaces need to render the provider chip on each
/// project-history row.
class ClientProjectProvider {
  const ClientProjectProvider({
    required this.organizationId,
    required this.displayName,
    this.avatarUrl,
  });

  /// Organization id of the provider that delivered this project.
  /// Used to link back to the provider's public profile.
  final String organizationId;

  /// Human-readable name rendered in the project row.
  final String displayName;

  /// Optional provider avatar URL.
  final String? avatarUrl;

  factory ClientProjectProvider.fromJson(Map<String, dynamic> json) {
    return ClientProjectProvider(
      organizationId: json['organization_id'] as String? ?? '',
      displayName: json['display_name'] as String? ?? '',
      avatarUrl: json['avatar_url'] as String?,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is ClientProjectProvider &&
        other.organizationId == organizationId &&
        other.displayName == displayName &&
        other.avatarUrl == avatarUrl;
  }

  @override
  int get hashCode =>
      Object.hash(organizationId, displayName, avatarUrl);
}

/// One completed project in the client's project history.
///
/// Amounts are stored in cents (the backend's canonical representation)
/// so consumers must format them before rendering. Each entry carries
/// the matching provider→client review inline when the provider has
/// already left feedback; otherwise [review] is null and the UI shows
/// an "Awaiting review" placeholder.
class ClientProjectEntry {
  const ClientProjectEntry({
    required this.proposalId,
    required this.title,
    required this.amount,
    required this.completedAt,
    required this.provider,
    this.review,
  });

  /// Unique id of the proposal that produced this project row.
  final String proposalId;

  /// Proposal title at completion time.
  final String title;

  /// Amount in cents. Caller formats for display (e.g. `€1234.56`).
  final int amount;

  /// Timestamp the proposal was marked completed.
  final DateTime completedAt;

  /// Provider side of the completed engagement.
  final ClientProjectProvider provider;

  /// Provider→client review for this project, when one has been
  /// submitted. `null` means the review is still pending.
  final Review? review;

  factory ClientProjectEntry.fromJson(Map<String, dynamic> json) {
    final rawReview = json['review'];
    return ClientProjectEntry(
      proposalId: json['proposal_id'] as String? ?? '',
      title: json['title'] as String? ?? '',
      amount: _readInt(json['amount']),
      completedAt: _parseDate(json['completed_at']) ?? _epoch,
      provider: ClientProjectProvider.fromJson(
        (json['provider'] as Map<String, dynamic>?) ?? const {},
      ),
      review: rawReview is Map<String, dynamic>
          ? Review.fromJson(rawReview)
          : null,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is ClientProjectEntry &&
        other.proposalId == proposalId &&
        other.title == title &&
        other.amount == amount &&
        other.completedAt == completedAt &&
        other.provider == provider &&
        other.review?.id == review?.id;
  }

  @override
  int get hashCode =>
      Object.hash(proposalId, title, amount, completedAt, provider, review?.id);
}

/// Public client profile aggregate — company identity + client-side
/// reputation + history of completed engagements.
///
/// Built around the locked contract for `GET /api/v1/clients/{orgId}`.
/// The private profile (editable screen) reuses the same fields by
/// cherry-picking values off the existing `GET /api/v1/profile`
/// response — this entity is the single source of truth regardless of
/// the fetch surface.
///
/// Reviews are no longer exposed as a top-level list — each
/// [ClientProjectEntry] carries the matching provider→client review
/// inline under [ClientProjectEntry.review].
class ClientProfile {
  const ClientProfile({
    required this.organizationId,
    required this.type,
    required this.companyName,
    required this.clientDescription,
    required this.totalSpent,
    required this.reviewCount,
    required this.averageRating,
    required this.projectsCompletedAsClient,
    this.avatarUrl,
    this.projectHistory = const <ClientProjectEntry>[],
  });

  /// Organization id — used to build the public profile URL.
  final String organizationId;

  /// Organization type. Expected to be `agency` or `enterprise`.
  /// `provider_personal` is rejected upstream (404).
  final String type;

  /// Public company name, shared with the provider profile.
  final String companyName;

  /// Optional company avatar URL.
  final String? avatarUrl;

  /// Free-form client-side description, client-only field (does not
  /// exist on the provider profile).
  final String clientDescription;

  /// Total amount spent as a client, in cents.
  final int totalSpent;

  /// Number of reviews received from providers.
  final int reviewCount;

  /// Average rating received from providers. 0 when no review.
  final double averageRating;

  /// Total number of proposals completed as a client.
  final int projectsCompletedAsClient;

  /// Completed projects, most recent first (the backend is responsible
  /// for ordering — we simply render the list as received).
  final List<ClientProjectEntry> projectHistory;

  factory ClientProfile.fromJson(Map<String, dynamic> json) {
    final rawHistory = json['project_history'];

    return ClientProfile(
      organizationId: json['organization_id'] as String? ?? '',
      type: json['type'] as String? ?? '',
      companyName: json['company_name'] as String? ?? '',
      avatarUrl: json['avatar_url'] as String?,
      clientDescription: json['client_description'] as String? ?? '',
      totalSpent: _readInt(json['total_spent']),
      reviewCount: _readInt(json['review_count']),
      averageRating: _readDouble(json['average_rating']),
      projectsCompletedAsClient:
          _readInt(json['projects_completed_as_client']),
      projectHistory: rawHistory is List
          ? rawHistory
              .whereType<Map<String, dynamic>>()
              .map(ClientProjectEntry.fromJson)
              .toList(growable: false)
          : const <ClientProjectEntry>[],
    );
  }

  /// Convenience — true when at least one review has been received.
  bool get hasReviews => reviewCount > 0 && averageRating > 0;
}

// ---------------------------------------------------------------------------
// Parsing helpers — defensive against missing / loosely-typed fields
// ---------------------------------------------------------------------------

final DateTime _epoch = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);

int _readInt(Object? value) {
  if (value == null) return 0;
  if (value is int) return value;
  if (value is double) return value.toInt();
  if (value is String) return int.tryParse(value) ?? 0;
  return 0;
}

double _readDouble(Object? value) {
  if (value == null) return 0.0;
  if (value is double) return value;
  if (value is int) return value.toDouble();
  if (value is String) return double.tryParse(value) ?? 0.0;
  return 0.0;
}

DateTime? _parseDate(Object? value) {
  if (value is String && value.isNotEmpty) {
    return DateTime.tryParse(value);
  }
  return null;
}

File: backend/internal/handler/dto/response/client_profile.go
package response

import (
	"time"

	"marketplace-backend/internal/app/clientprofile"
)

// ClientProjectHistoryProvider is the compact counterparty descriptor
// embedded in each client-side project history entry. Surfaces only
// the public identity bits (org id + display name + avatar) — private
// profile fields never leak onto this surface.
type ClientProjectHistoryProvider struct {
	OrganizationID string `json:"organization_id"`
	DisplayName    string `json:"display_name"`
	AvatarURL      string `json:"avatar_url"`
}

// ClientProjectHistoryEntry is the public API shape of one completed
// deal where the org was the client. Title is empty when the provider
// did not elect to share it in their review.
type ClientProjectHistoryEntry struct {
	ProposalID  string                        `json:"proposal_id"`
	Title       string                        `json:"title"`
	Amount      int64                         `json:"amount"`
	CompletedAt string                        `json:"completed_at"`
	Provider    *ClientProjectHistoryProvider `json:"provider"`
}

// PublicClientProfileResponse is the public read shape for the
// /api/v1/clients/{orgId} endpoint. Every slice and every nullable is
// explicitly modeled so the JSON shape is stable across requests.
//
// Reviews are NOT surfaced as a top-level list — the review attached
// to each completed deal travels inline on project_history[].review,
// and the frontend renders one unified "Completed projects" section
// (mirroring the provider profile). ReviewCount + AverageRating stay
// at the top level because they feed the header stats block.
type PublicClientProfileResponse struct {
	OrganizationID            string                      `json:"organization_id"`
	Type                      string                      `json:"type"`
	CompanyName               string                      `json:"company_name"`
	AvatarURL                 string                      `json:"avatar_url"`
	ClientDescription         string                      `json:"client_description"`
	TotalSpent                int64                       `json:"total_spent"`
	ReviewCount               int                         `json:"review_count"`
	AverageRating             float64                     `json:"average_rating"`
	ProjectsCompletedAsClient int                         `json:"projects_completed_as_client"`
	ProjectHistory            []ClientProjectHistoryEntry `json:"project_history"`
}

// NewPublicClientProfileResponse builds the public response envelope
// from the app-layer aggregate. Every slice is pre-allocated (non-nil
// even when empty) so the JSON output does not alternate between []
// and null across requests.
func NewPublicClientProfileResponse(p *clientprofile.PublicClientProfile) PublicClientProfileResponse {
	history := make([]ClientProjectHistoryEntry, 0, len(p.ProjectHistory))
	for _, e := range p.ProjectHistory {
		history = append(history, newClientProjectHistoryEntry(e))
	}
	return PublicClientProfileResponse{
		OrganizationID:            p.OrganizationID.String(),
		Type:                      p.Type,
		CompanyName:               p.CompanyName,
		AvatarURL:                 p.AvatarURL,
		ClientDescription:         p.ClientDescription,
		TotalSpent:                p.TotalSpent,
		ReviewCount:               p.ReviewCount,
		AverageRating:             p.AverageRating,
		ProjectsCompletedAsClient: p.ProjectsCompletedAsClient,
		ProjectHistory:            history,
	}
}

// newClientProjectHistoryEntry maps one app-layer entry to the API
// shape. Nil provider stays nil so the frontend can render a generic
// placeholder for deleted counterparties without a defensive
// guard on other fields.
func newClientProjectHistoryEntry(e clientprofile.ProjectHistoryEntry) ClientProjectHistoryEntry {
	entry := ClientProjectHistoryEntry{
		ProposalID:  e.ProposalID.String(),
		Title:       e.Title,
		Amount:      e.Amount,
		CompletedAt: e.CompletedAt.Format(time.RFC3339),
	}
	if e.Provider != nil {
		entry.Provider = &ClientProjectHistoryProvider{
			OrganizationID: e.Provider.OrganizationID.String(),
			DisplayName:    e.Provider.Name,
			AvatarURL:      e.Provider.PhotoURL,
		}
	}
	return entry
}

// NewProfileClientSection maps the client-stats aggregate to the DTO
// used inline by GET /api/v1/profile. Nil input returns nil so the
// caller can safely chain into ProfileResponse.WithClientSection.
func NewProfileClientSection(stats *clientprofile.ClientStats) *ProfileClientSection {
	if stats == nil {
		return nil
	}
	return &ProfileClientSection{
		TotalSpent:                stats.TotalSpent,
		ReviewCount:               stats.ReviewCount,
		AverageRating:             stats.AverageRating,
		ProjectsCompletedAsClient: stats.ProjectsCompletedAsClient,
	}
}
